### PR TITLE
[Dispatchers] Refactor architecture for clarity

### LINKED ISSUE #2968: [Dispatcher] internal fields misleadingly named
In `Dispatcher` there is this field:
```Typescript
protected _eventToCallback: { [eventName: string]: (e: Event) => any; } = {};
```

But also these:
```Typescript
protected _callbacks: Utils.CallbackSet<Function>[] = [];

protected _setCallback(callbackSet: Utils.CallbackSet<Function>, callback: Function);
protected _unsetCallback(callbackSet: Utils.CallbackSet<Function>, callback: Function);
```

The naming here is misleading since there are actually *two* kinds of callbacks involved.`_eventToCallback` is closer to "event to processing-function-*of*-`Dispatcher`", and `_set`/`_unsetCallback` is closer to "set/unset callbacks to be *invoked by* `Dispatcher`.

Proposed renamings:
`_eventToCallback` --> `_eventToProcessingFunction`
`_callbacks` --> `_callbackSets`
`_setCallback` --> `_addCallback`
`_unsetCallback` --> `_removeCallback`

It's also likely that `_callbackSets` should be privatized and the signatures of `_set`/`_unsetCallback` changed:
```Typescript
private _callbackSets: { [eventName: string]: Utils.CallbackSet<Function>; }

protected _setCallback(eventName: string, callback: Function);
protected _unsetCallback(eventName: string, callback: Function);
```

This seems more consistent.

### PR DESCRIPTION
**Pull request by jtlan**
Refactored how `Dispatcher` registers events. Previously it used these methods:
```Typescript
// OLD DESIGN
protected _setCallback(callbackSet: Utils.CallbackSet<Function>, callback: Function): void;
protected _unsetCallback(callbackSet: Utils.CallbackSet<Function>, callback: Function): void;
```
This seemed kind of odd, since there was a method that modified non-instance fields, and subclasses had to define a new `CallbackSet` for each type of event they wanted to add listeners for, and then call `.callCallbacks()` manually. This seemed silly.

The refactored version:
```Typescript
// NEW DESIGN
protected _addCallbackForEvent(eventName: string, callback: Function): void;
protected _removeCallbackForEvent(eventName: string, callback: Function): void;
protected _callCallbacksForEvent(eventName: string, ...args: any[]): void;
```
Now, subclasses register callbacks using a unique event-name, and the root `Dispatcher` class handles creating `CallbackSet`s. This change also meant we could remove the `protected` (and [misleadingly named](https://github.com/palantir/plottable/issues/2968)!) `_callbacks` array.

Summary of changes:
* ` _eventToCallback` --> `_eventToProcessingFunction`
* `_setCallback()` and `_unsetCallback()` replaced by `_addCallbackForEvent()` and `_removeCallbackForEvent()`.
* `_callCallbacksForEvent()` added.
* `protected` `_callbacks` array removed.
* `private` `CallbackSet`s removed and replaced with registration-strings.

Close #2968.

### CODE REVIEW

File: src/dispatchers/mouseDispatcher.ts [hunk 1]
```diff
@@ -183,7 +180,7 @@ module Plottable.Dispatchers {
         let newMousePosition = this._translator.computePosition(event.clientX, event.clientY);
         if (newMousePosition != null) {
           this._lastMousePosition = newMousePosition;
-          callbackSet.callCallbacks(this.lastMousePosition(), event);
+          this._callCallbacksForEvent(eventName, this._lastMousePosition, event);
```
Comment (ztsai) [inline src/dispatchers/mouseDispatcher.ts]: either use `this.lastMousePosition()` or remove that function

File: src/dispatchers/touchDispatcher.ts [hunk 2]
```diff
@@ -38,16 +38,14 @@ module Plottable.Dispatchers {
...
-      this._moveCallbacks = new Utils.CallbackSet<TouchCallback>();
-      this._endCallbacks = new Utils.CallbackSet<TouchCallback>();
-      this._cancelCallbacks = new Utils.CallbackSet<TouchCallback>();
-      this._callbacks = [this._moveCallbacks, this._startCallbacks, this._endCallbacks, this._cancelCallbacks];
-
-      this._eventToCallback["touchstart"] = (e: TouchEvent) => this._measureAndDispatch(e, this._startCallbacks, "page");
-      this._eventToCallback["touchmove"] = (e: TouchEvent) => this._measureAndDispatch(e, this._moveCallbacks, "page");
-      this._eventToCallback["touchend"] = (e: TouchEvent) => this._measureAndDispatch(e, this._endCallbacks, "page");
-      this._eventToCallback["touchcancel"] = (e: TouchEvent) => this._measureAndDispatch(e, this._cancelCallbacks, "page");
+      this._eventToProcessingFunction[Touch._TOUCHSTART_EVENT_NAME] =
```
Comment (ztsai) [inline src/dispatchers/touchDispatcher.ts]: probably in a different commit, but all these events are `page` events so do we still need that extra parameter?

File: src/dispatchers/mouseDispatcher.ts (hunk 1 above)
Comment (jtlan) [inline src/dispatchers/mouseDispatcher.ts]: ... strange, I don't remember changing this, but I'll put it back the way it was.

File: src/dispatchers/touchDispatcher.ts (hunk 2 above)
Comment (jtlan) [inline src/dispatchers/touchDispatcher.ts]: Which extra parameter? `_measureAndDispatch()` defaults to `"element"` so we need to pass `"page"`.

File: src/dispatchers/dispatcher.ts [hunk 3]
```diff
@@ -1,42 +1,60 @@
...
   }
 
   private _connect() {
     if (this._connected) {
       return;
     }
-    Object.keys(this._eventToCallback).forEach((event: string) => {
-      let callback = this._eventToCallback[event];
+    Object.keys(this._eventToProcessingFunction).forEach((event: string) => {
+      let callback = this._eventToProcessingFunction[event];
```
Comment (ztsai) [inline src/dispatchers/dispatcher.ts]: nitpicking, but processingFunction is a better variable name here

File: src/dispatchers/dispatcher.ts [hunk 4]
```diff
@@ -1,42 +1,60 @@
...
     this._connected = true;
   }
 
   private _disconnect() {
-    if (this._connected && this._hasNoListeners()) {
-      Object.keys(this._eventToCallback).forEach((event: string) => {
-        let callback = this._eventToCallback[event];
+    if (this._connected && this._hasNoCallbacks()) {
+      Object.keys(this._eventToProcessingFunction).forEach((event: string) => {
+        let callback = this._eventToProcessingFunction[event];
```
Comment (ztsai) [inline src/dispatchers/dispatcher.ts]: same here

File: src/dispatchers/touchDispatcher.ts (hunk 2 above)
Comment (ztsai) [inline src/dispatchers/touchDispatcher.ts]: I don't think `scope` parameter is necessary in  `_measureAndDispatch ` as for current implementation, it's doing additional check for nothing; but I am also okay with leaving it there to be consistent with the Mouse Dispatcher

Comment (jtlan) [inline src/dispatchers/touchDispatcher.ts]: Seems like a refactor for later.

**Pull request merged**